test/e2e/kcp_remediations.go

```diff
@@ -0,0 +1,693 @@
+			expectedDeletedMachines:         []string{thirdMachineName},
+			expectedOldMachines:             []string{firstMachineReplacementName, secondMachineReplacementName},
+			waitForMachinesIntervals:        input.E2EConfig.GetIntervals(specName, "wait-machines"),
+			checkMachineListStableIntervals: input.E2EConfig.GetIntervals(specName, "check-machines-stable"),
+		})
+		Expect(allMachines).To(HaveLen(3))
+		Expect(newMachines).To(HaveLen(1))
+		thirdMachineReplacementName := newMachines[0]
+		thirdMachineReplacement := &clusterv1.Machine{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      thirdMachineReplacementName,
+				Namespace: namespace.Name,
+			},
+		}
+		Expect(input.BootstrapClusterProxy.GetClient().Get(ctx, client.ObjectKeyFromObject(thirdMachineReplacement), thirdMachineReplacement)).To(Succeed(), "Failed to get machine %d", thirdMachineReplacementName)
+		Expect(thirdMachineReplacement.Status.NodeRef).To(BeNil())
+		log.Logf("Machine %s is up but still bootstrapping", thirdMachineReplacementName)
+
+		// The thirdMachine replacement is up, meaning that the test validated that remediation of the third CP machine works (note: this test remediation after the cluster is fully provisioned).
+
+		By("THIRD CP SUCCESSFULLY REMEDIATED!")
+
+		Byf("Unblock bootstrap for Machine %s and wait for it to be provisioned", thirdMachineReplacementName)
+		sendSignalToBootstrappingMachine(ctx, sendSignalToBootstrappingMachineInput{
+			client:    input.BootstrapClusterProxy.GetClient(),
+			namespace: namespace.Name,
+			machine:   thirdMachineReplacementName,
+			signal:    "pass",
+		})
+		log.Logf("Waiting for Machine %s to be provisioned", thirdMachineReplacementName)
+		Eventually(func() bool {
+			if err := input.BootstrapClusterProxy.GetClient().Get(ctx, client.ObjectKeyFromObject(thirdMachineReplacement), thirdMachineReplacement); err != nil {
+				return false
+			}
+			return thirdMachineReplacement.Status.NodeRef != nil
+		}, input.E2EConfig.GetIntervals(specName, "wait-machine-provisioned")...).Should(BeTrue(), "Machine %s failed to be provisioned", thirdMachineReplacementName)
+
+		// All three CP machines are up again.
+
+		By("CP BACK TO FULL OPERATIONAL STATE!")
+
+		By("PASSED!")
+	})
+
+	AfterEach(func() {
+		// Dumps all the resources in the spec namespace, then cleanups the cluster object and the spec namespace itself.
+		dumpSpecResourcesAndCleanup(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder, namespace, cancelWatches, clusterResources.Cluster, input.E2EConfig.GetIntervals, input.SkipCleanup)
+	})
+}
+
+func createConfigMapForMachinesBootstrapSignal(ctx context.Context, writer client.Writer, namespace string) {
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      configMapName,
+			Namespace: namespace,
+		},
+		Data: map[string]string{
+			configMapDataKey: "hold",
+		},
+	}
+	Expect(writer.Create(ctx, cm)).To(Succeed(), "failed to create mhc-test config map")
```

Comment:
Wondering if we should extend the scope of this func a bit. Essentially `applyConfigMapForMachinesBootstrapSignal` which does an create or patch and the signal can be passed in. Could be used to remove some low-level code from `sendSignalToBootstrappingMachine`

Reply:
Frankly speaking, I don't think it is worth the effort, we need to maintain both the creation and the update path, no matter if we shuffle them from one func to another, and upgrade happens only during sendSignalToBootstrappingMachine.